src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs

```diff
@@ -0,0 +1,146 @@
+            IQueryResultRetriever retriever, DocumentsOperationContext documentsContext, Func<string, SpatialField> getSpatialField, CancellationToken token)
+        {
+            var pageSize = query.PageSize;
+            var isDistinctCount = pageSize == 0 && query.Metadata.IsDistinct;
+            if (isDistinctCount)
+                pageSize = int.MaxValue;
+            var position = query.Start;
+            long readCounter = 0;
+
+            IQueryMatch result = _coraxQueryEvaluator.Search(query.Metadata, fieldsToFetch);
+            if (result == null)
+                yield break;
+
+            totalResults.Value = Convert.ToInt32(result.Count);
+            var ids = ArrayPool<long>.Shared.Rent(_bufferSize);
+            try
+            {
+                int read = 0;
+
+                //escaping already returned docs.
+                // Q: Can Corax have "skip" method inside? It's inefficient to copy values only for skipping.
+                int docsToLoad = pageSize;
+                if (position != 0)
+                {
+                    int emptyRead = position / _bufferSize;
+                    do
+                    {
+                        read = result.Fill(ids);
+                        readCounter += read;
+                        emptyRead--;
+                    } while (emptyRead > 0);
+
+                    position %= _bufferSize; // move into <0;_bufferSize> set.
+                    //I know there is a cost of copying this but it's max _bufferSize and make code much simpler.
+                    ids[position..read].CopyTo(ids, 0);
+                    read -= position;
+                    skippedResults.Value = -1; // docsToLoad * _bufferSize + position;
+                }
+
+                // first Fill operation would be done outside loop because we need to check if there is already some data read.
+                if (read == 0)
+                {
+                    read = result.Fill(ids);
+                    readCounter += read;
+                }
+
+                while (read != 0)
+                {
+                    for (int i = 0; i < read && docsToLoad != 0; --docsToLoad, ++i)
+                    {
+                        RetrieverInput retrieverInput = new(_indexSearcher.GetReaderFor(ids[i]), _indexSearcher.GetIdentityFor(ids[i]));
+
+                        yield return new QueryResult() { Result = retriever.Get(ref retrieverInput) };
+                    }
+
+                    if (docsToLoad == 0)
+                        break;
+
+                    read = result.Fill(ids);
+                    readCounter += read;
+                }
+
+                //special value for studio. We have unbounded set and don't know how much items is in index so it should count it during loading.
+                if (totalResults.Value == 0)
+                {
+                    totalResults.Value = Convert.ToInt32(readCounter);
```

Comment:
That isn't actually correct? You read the first X values, but we know what the _total_ number is.
I'm not sure about the Studio reference here. 

@redknightlois I think that we need to expose the total number of query matches here, so we can report that to the user.
Note that in any case where we have an order by, we'll have to go through the whole list, so it shouldn't be complex to do so.

The value should be `long`

Reply:
For query:
`from Orders where startsWith(Company, "comp")`
![image](https://user-images.githubusercontent.com/86351904/132014118-21211083-9387-4ade-89d8-d6a670510690.png)
Maybe there is a bug? @redknightlois ?
Having a count would be great. :)

Reply:
@redknightlois can you have a look here?

Reply:
We talked today about this particular one. That count has a different use, it is coupled to the `QueryCountConfidence` it shouldnt be treated as a reliable value. This is used only for query planning purposes. If you need to know the exact amount, you need to execute the query.

Reply:
Note that at this point, we *have* executed the query. But we don't provide a way to tell what the total results were.

Reply:
Should I iterate through all elements doing the first query to get total results and save it in cache?

Reply:
No, you can't do that. We are already doing filtering on that during the sorting match. I'll fix that shortly.

Reply:
https://github.com/ravendb/ravendb/commit/96c2ce2d106b9e7c5e36759140ad3d8ce9a666d5